man/genbf.1
.TH genbf 1 "31 July 1995" "GIG's Utilities" "Version 1.05"
.SH NAME
genbf - GENerate a Binary File.
.SH SYNOPSIS
genbf [flags] filename
.SH DESCRIPTION
Writes the specified file with a set of patterns
of characters suitable for testing other
file manipulation utilities.

This utility was written to generate files
on which to test other utilities such as
DISF, COMF, and ZAPF.
To do so, these test files needed to contain
all possible character codes from 0 to 255,
in ascending ("forward") and descending ("reverse") order.
In addition, they needed to contain blocks of identical characters.
The test file is generated based on the values of
command line flags.
The -R flag controls the number of "reverse" blocks,
while -F controls the number of "forward" blocks.
Blocks of identical characters are controlled by the
-B, -C, and -L flags which specify
the number, character code, and size of each block,
respectively.

The flags are processed in the order -R, -B/-C/-L, -F.
That is, reverse blocks are written first,
custom or user defined blocks are written second,
and forward blocks are written last.
If no flags are specified to indicate the number
and types of blocks to write to the specified file,
then a default set of flags is used:

   -R1 -B2 -C33 -L256 -F1

That is,
one 256 byte reverse block
is followed by
two 256 byte blocks containing "!" characters (decimal 33)
and these are followed by
one 256 byte forward block,
for a total of 1024 bytes written. 
.SH PARAMETERS
.TP 12
-Rn
Specifies that <n> blocks of all characters
in reverse order (0xFF through 0x00)
are to be written to the file.
.TP
-Bn
Specifies that <n> user specified blocks
are to be written to the file.
The size of these blocks are specified by the -L switch,
and the contents of these blocks are specified by the -C
switch.
.TP
-Cn
Specifies that the user specified blocks
are to contain the character whose decimal value is <n>
replicated throughout the block.
Therefore, the value of <n> must be
the value of an unsigned character,
between 0 and 255, inclusive.
.TP
-Ln
Specifies that the user specified blocks
are to be <n> bytes long.
The value of <n> must be greater than one (1),
and is arbitrarily limited by the size of a
two byte signed integer, or 32767.
.TP
-Fn
Specifies that <n> blocks of all characters
in ascending ("forward") order (0x00 through 0xFF)
are to be written to the file.
.TP
-V
Specifies that information concerning the
internal operation of the utility is to be
written to stdout.
This "verbose" output is normally used to debug the utility.
.SH EXAMPLES
To generate a simple binary test file, use the default switches.

   % genbf xxx
   Generate binary file named "xxx".
      No block types specified.   Using -R1 -B2 -C33 -L256 -F1.

This creates a 1024 byte file described above.
A simpler file containing all 256 possible bytes
can be generated by specifying one switch.

   % genbf -f xxx
   Generate binary file named "xxx".

This creates a 256 byte file containing one "forward" block,
that is, all bytes starting with 0x00 through 0xFF.

GENBF has other uses than just creating test files.
For example, the following command line can be used on a PC
to scrub a 3.5" HD diskette one track at a time.

   C:\\>GENBF A:X -L9216 -B158 -C33

It writes 158 blocks, the number of data tracks,
that are 9216 bytes long, the size of each track,
filled with character 33 (0x21 is bang!).
.SH "SEE ALSO"
comf(1),
cvtbt(1),
cvttb(1),
disf(1),
zapf(1).
.SH BUGS
Due to the mechanism by which command line switches are processed,
there currently isn't a simple way to control the order in which
blocks are written to the binary file by simple arrangement of the
command line switches.

It would be nice to be able to specify more complex blocks than
just simple chunks of the same character.
Patterns other than all bytes ascending and descending would be
nice, too.
.SH AUTHOR
Greg Grimes <sasgig@unx.sas.com>


src/genbf.c
#define   VERSION   "05"
#include "us.h"
#include "usargs.h"
#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#ifndef _NOSTDLIB
#include <stdlib.h>
#endif

   static char buffer[32768];

int main(argc,argv)
 int argc;
 char *argv[];
{
   int   i;    
   int   n;
   FILE *file;
   char *arg;
   char *fname;

/**********************************************************************/
/* Process command line arguments and get file name                   */
/**********************************************************************/
   arg=USARGS();

/**********************************************************************/
/* Show help, if requested                                            */
/**********************************************************************/
   if(FLAG(?) || arg==NULL) 
      {
      US_NAME(arg)
      printf("Usage:\n");
      printf("   %s [flags] filename\n",arg);
      printf("\n");
      printf("   Builds a file with various binary structures.\n");
      printf("\n");
      printf("flags:\n");
      printf("   -Rn  include n blocks containing 0xFF through 0x00\n");
      printf("   -Bn  include n \"custom\" blocks\n");
      printf("   -Cn  custom blocks use byte value n (0<=n<=255)\n");
      printf("   -Ln  custom blocks are n bytes long (1<=n<=32767)\n");
      printf("   -Fn  include n blocks containing 0x00 through 0xFF\n");  
      printf(US_FLAGV);
      printf("\n");
      printf(US_COPYRIGHT);
      exit(0);
      }

/**********************************************************************/
/* Report file name to be generated                                   */
/**********************************************************************/
   fname = arg;
#ifdef DOS
   while(*arg) {*arg=toupper(*arg);arg++;}
#endif
   printf("Generate binary file named \"%s\".\n",fname);

/**********************************************************************/
/* Check for something to write                                       */
/**********************************************************************/
   if(FLAG(B)==0 && FLAG(C)==0 && FLAG(L)==0 && 
      FLAG(R)==0 && FLAG(F)==0)
      {
      printf("   No block types specified.");
      printf("   Using -R1 -B2 -C33 -L256 -F1.\n");
      FLAG(R) = 1;
      FLAG(B) = 2;
      FLAG(C) = 33;
      FLAG(L) = 256;
      FLAG(F) = 1;
      }

/**********************************************************************/
/* Validate Block Values                                              */
/**********************************************************************/
   if(FLAG(R) < 0)
      {
      printf("   Instead of -R%d, using -R1.\n",FLAG(R));
      FLAG(R) = 1;
      }

   if(FLAG(B)!=0 || FLAG(C)!=0 || FLAG(L)!=0)
      {
      if(FLAG(B) < 1)
         {
         printf("   Instead of -B%d, using -B1.\n",FLAG(B));
         FLAG(B) = 1;
         }
      if(FLAG(C) < 0  ||  FLAG(C) > 255)
         {
         printf("   Instead of -C%d (0x%2.2X = '%c'),\n",
            FLAG(C),FLAG(C),FLAG(C));
         FLAG(C) = '!';
         printf("        using -C%d (0x%2.2X = '%c').\n",
            FLAG(C),FLAG(C),FLAG(C));
         }
      if(FLAG(L) < 1  ||  FLAG(L) > 32767)
         {
         printf("   Instead of -L%d, using -L256.\n",FLAG(L));
         FLAG(L) = 256;
         }
      }

   if(FLAG(F) < 0)
      {
      printf("   Instead of -F%d, using -F1.\n",FLAG(F));
      FLAG(F) = 1;
      }

/**********************************************************************/
/* Debug information                                                  */
/**********************************************************************/
   if(FLAG(V))
      {
      if(FLAG(R))
         printf("   %2d Reverse block(s).\n",FLAG(R));
      else
         printf("   No Reverse blocks.\n");

      if(FLAG(B))
         {
         printf("   %2d Custom  block(s) of %d bytes",FLAG(B),FLAG(L));
         printf(" with value %d (0x%2.2X = '%c').\n",FLAG(C),FLAG(C),FLAG(C));
         }
      else
         printf("   No Custom  blocks.\n");

      if(FLAG(F))
         printf("   %2d Forward block(s).\n",FLAG(F));
      else
         printf("   No Forward blocks.\n");
      }

/**********************************************************************/
/* Open the file to be generated                                      */
/**********************************************************************/
   file = fopen(fname,"wb");
   if(file == NULL)
      {
      printf("Unable to open %s for output, errno=%d (0x%0X).\n",
         fname,errno,errno);
      exit(1);
      }
   if(FLAG(V))
      printf("   File opened with file pointer %p, errno=%d (0x%0X).\n",
         file,errno,errno);

/**********************************************************************/
/* Build and write "Reverse" blocks                                   */
/**********************************************************************/
   if(FLAG(R))
      {
      if(FLAG(V))  printf("   Building Reverse buffer.\n");
      for(i=0;i<256;i++)
         buffer[i]=255-i;
      if(FLAG(V))  printf("   Writing %d buffer images.\n",FLAG(R));
      for(n=0;n<FLAG(R);n++)
         fwrite(buffer,256,1,file);
      }

/**********************************************************************/
/* Build and write "Custom" blocks                                    */
/**********************************************************************/
   if(FLAG(B))
      {
      if(FLAG(V))  printf("   Building Custom buffer.\n");
      for(i=0;i<FLAG(L);i++)
         buffer[i]=FLAG(C);
      if(FLAG(V))  printf("   Writing %d buffer images.\n",FLAG(B));
      for(n=0;n<FLAG(B);n++)
         fwrite(buffer,FLAG(L),1,file);
      }

/**********************************************************************/
/* Build and write "Forward" blocks                                   */
/**********************************************************************/
   if(FLAG(F))
      {
      if(FLAG(V))  printf("   Building Forward buffer.\n");
      for(i=0;i<256;i++)
         buffer[i]=i;
      if(FLAG(V))  printf("   Writing %d buffer images.\n",FLAG(F));
      for(n=0;n<FLAG(F);n++)
         fwrite(buffer,256,1,file);
      }

/**********************************************************************/
/* Close the generated file                                           */
/**********************************************************************/
   i = fclose(file);
   if(FLAG(V))
      printf("   File closed with rc=%d (0x%0X), errno=%d (0x%0X).\n",
         i,i,errno,errno);

/**********************************************************************/
/* Return                                                             */
/**********************************************************************/
   return 0;
}
